src/libraries/Microsoft.PowerFx.LanguageServerProtocol/LanguageServer/LanguageServer.cs

```diff
@@ -690,8 +691,36 @@ private void HandleSemanticTokens<T>(string id, T semanticTokensParams, string m
                 tokens = tokens.Where(token => !(token.EndIndex <= startIndex || token.StartIndex >= endIndex));
             }
 
-            var encodedTokens = SemanticTokensEncoder.EncodeTokens(tokens, expression, eol);
+            var controlTokens = isFullDocument ? new LinkedList<ControlToken>() : null;
+
+            var encodedTokens = SemanticTokensEncoder.EncodeTokens(tokens, expression, eol, controlTokens);
             _sendToClient(JsonRpcHelper.CreateSuccessResult(id, new SemanticTokensResponse() { Data = encodedTokens }));
+
+            PublishControlTokenNotification(controlTokens, queryParams);
+        }
+
+        /// <summary>
+        /// Handles publishing a control token notification if any control tokens found.
+        /// </summary>
+        /// <param name="controlTokens">Collection to add control tokens to.</param>
+        /// <param name="queryParams">Collection of query params.</param>
+        private void PublishControlTokenNotification(IEnumerable<ControlToken> controlTokens, NameValueCollection queryParams)
+        {
+            if (controlTokens == null || controlTokens.Count() == 0 || queryParams == null)
+            {
+                return;
+            }
+
+            var version = queryParams.Get("version") ?? string.Empty;
+
+            // Send PublishControlTokens notification
+            _sendToClient(JsonRpcHelper.CreateNotification(
+                CustomProtocolNames.PublishControlTokens,
+                new PublishControlTokensParams()
+                {
+                    Version = version,
+                    Data = controlTokens
```

Comment:
Name this as Controls and not Data. Semantic tokens is general while this is specific.